server/src/main/java/org/elasticsearch/index/search/MultiMatchQueryParser.java

```diff
@@ -106,15 +99,216 @@ private Query combineGrouped(List<Query> groupQuery, float tieBreaker) {
+    }
+
+    private class CombinedFieldsQueryBuilder extends MatchQueryBuilder {
+        private final List<FieldAndBoost> fields;
+
+        CombinedFieldsQueryBuilder(Analyzer analyzer, List<FieldAndBoost> fields,
+                                   boolean enablePositionIncrements, boolean autoGenerateSynonymsPhraseQuery) {
+            super(analyzer, fields.get(0).fieldType, enablePositionIncrements, autoGenerateSynonymsPhraseQuery);
+            this.fields = fields;
+        }
+
+        @Override
+        public Query createPhraseQuery(String field, String queryText, int phraseSlop) {
+            throw new IllegalArgumentException("[multi_match] queries in [combined_fields] mode don't support phrases");
+        }
+
+        @Override
+        protected Query createPhrasePrefixQuery(String field, String queryText, int slop) {
+            throw new IllegalArgumentException("[multi_match] queries in [combined_fields] mode don't support phrase prefix");
+        }
+
+        @Override
+        protected Query createBooleanPrefixQuery(String field, String queryText, BooleanClause.Occur occur) {
+            throw new IllegalArgumentException("[multi_match] queries in [combined_fields] mode don't support boolean prefix");
+        }
+
+        @Override
+        protected Query newSynonymQuery(TermAndBoost[] terms) {
+            List<FieldAndBoost> textFields = new ArrayList<>();
+            List<FieldAndBoost> nonTextFields = new ArrayList<>();
+
+            for (FieldAndBoost fieldAndBoost : fields) {
+                MappedFieldType fieldType = fieldAndBoost.fieldType;
+                if (fieldType.getTextSearchInfo().hasTerms()) {
+                    textFields.add(fieldAndBoost);
+                } else {
+                    nonTextFields.add(fieldAndBoost);
+                }
+            }
+
+            Query textQuery = createTextQuery(terms, textFields);
+            if (nonTextFields.isEmpty()) {
+                return textQuery;
+            } else {
+                BooleanQuery.Builder builder = new BooleanQuery.Builder();
+                builder.add(textQuery, BooleanClause.Occur.SHOULD);
+                addNonTextQueries(terms, nonTextFields, builder);
+                return builder.build();
+            }
+        }
+
+        private XCombinedFieldQuery createTextQuery(TermAndBoost[] terms, List<FieldAndBoost> fields) {
+            XCombinedFieldQuery.Builder query = new XCombinedFieldQuery.Builder();
+            for (TermAndBoost termAndBoost : terms) {
+                assert termAndBoost.boost == BoostAttribute.DEFAULT_BOOST;
+                BytesRef bytes = termAndBoost.term.bytes();
+                query.addTerm(bytes);
+            }
+            for (FieldAndBoost fieldAndBoost : fields) {
+                MappedFieldType fieldType = fieldAndBoost.fieldType;
+                float fieldBoost = fieldAndBoost.boost;
+                if (fieldType.getTextSearchInfo().hasTerms()) {
```

Comment:
nit: It seems that this check of `hasTerms()` is redundant, as we are only iterating through textFields on which we already had this check previously.